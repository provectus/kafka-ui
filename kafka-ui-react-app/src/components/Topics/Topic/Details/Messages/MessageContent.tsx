import React from 'react';

import FullMessage from './FullMessage';

interface MessageContentProps {
  message: string;
}

const MessageContent: React.FC<MessageContentProps> = ({ message }) => {
  const [isFolded, setIsFolded] = React.useState(message.length > 250);

  return (
    <div className="is-flex is-flex-direction-column">
      {isFolded ? (
        <p className="has-content-overflow-ellipsis">
          {`${message.slice(0, 250)}...`}
        </p>
      ) : (
        <FullMessage message={message} />
      )}
      {isFolded && (
        <button
          type="button"
          className="button is-small mt-2"
          onClick={() => setIsFolded((state) => !state)}
          title="Expand to JSON"
        >
          <span className="icon is-small">
            <i className="fas fa-chevron-down" />
          </span>
        </button>
      )}
    </div>
  );
};

export default MessageContent;
