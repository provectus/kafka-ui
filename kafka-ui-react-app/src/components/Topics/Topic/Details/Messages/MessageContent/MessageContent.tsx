import { TopicMessageTimestampTypeEnum } from 'generated-sources';
import React from 'react';
import JSONViewer from 'components/common/JSONViewer/JSONViewer';
import SecondaryTabsStyles from 'components/common/Tabs/SecondaryTabs.styled';
import { isObject } from 'lodash';
import BytesFormatted from 'components/common/BytesFormatted/BytesFormatted';

import {
  ContentBox,
  JSONViewerWrapper,
  MessageContentWrapper,
  Metadata,
  MetadataLabel,
  MetadataMeta,
  MetadataValue,
  MetadataWrapper,
} from './MessageContent.styled';

type Tab = 'key' | 'content' | 'headers';

export interface MessageContentProps {
  messageKey?: string;
  messageContent?: string;
  headers?: { [key: string]: string | undefined };
  timestamp?: Date;
  timestampType?: TopicMessageTimestampTypeEnum;
}

const MessageContent: React.FC<MessageContentProps> = ({
  messageKey,
  messageContent,
  headers,
  timestamp,
  timestampType,
}) => {
  const [activeTab, setActiveTab] = React.useState<Tab>('content');
  const activeTabContent = () => {
    switch (activeTab) {
      case 'content':
        return messageContent;
      case 'key':
        return messageKey;
      default:
        return JSON.stringify(headers);
    }
  };
  const handleKeyTabClick = (e: React.MouseEvent) => {
    e.preventDefault();
    setActiveTab('key');
  };
  const handleContentTabClick = (e: React.MouseEvent) => {
    e.preventDefault();
    setActiveTab('content');
  };
  const handleHeadersTabClick = (e: React.MouseEvent) => {
    e.preventDefault();
    setActiveTab('headers');
  };
  const keySize = new TextEncoder().encode(messageKey).length;
  const contentSize = new TextEncoder().encode(messageContent).length;
  const isContentJson = () => {
    try {
      return isObject(messageContent && JSON.parse(messageContent));
    } catch {
      return false;
    }
  };
  const isKeyJson = () => {
    try {
      return isObject(messageKey && JSON.parse(messageKey));
    } catch {
      return false;
    }
  };
  return (
    <MessageContentWrapper>
      <td colSpan={5}>
        <ContentBox>
          <SecondaryTabsStyles>
            <button
              type="button"
              className={activeTab === 'key' ? 'is-active' : ''}
              onClick={handleKeyTabClick}
            >
              Key
            </button>
            <button
              className={activeTab === 'content' ? 'is-active' : ''}
              type="button"
              onClick={handleContentTabClick}
            >
              Content
            </button>
            <button
              className={activeTab === 'headers' ? 'is-active' : ''}
              type="button"
              onClick={handleHeadersTabClick}
            >
              Headers
            </button>
          </SecondaryTabsStyles>
          <JSONViewerWrapper>
            <JSONViewer data={activeTabContent() || ''} />
          </JSONViewerWrapper>
        </ContentBox>
      </td>
      <td colSpan={2}>
        <MetadataWrapper>
          <Metadata>
            <MetadataLabel>Timestamp</MetadataLabel>
            <span>
              <MetadataValue>{timestamp?.toLocaleString()}</MetadataValue>
              <MetadataMeta>Timestamp type: {timestampType}</MetadataMeta>
            </span>
          </Metadata>

          <Metadata>
            <MetadataLabel>Content</MetadataLabel>
            <span>
              <MetadataValue>{isContentJson() ? 'JSON' : 'Text'}</MetadataValue>
              <MetadataMeta>
                Size: <BytesFormatted value={contentSize} />
              </MetadataMeta>
            </span>
          </Metadata>

          <Metadata>
            <MetadataLabel>Key</MetadataLabel>
            <span>
              <MetadataValue>{isKeyJson() ? 'JSON' : 'Text'}</MetadataValue>
              <MetadataMeta>
                Size: <BytesFormatted value={keySize} />
              </MetadataMeta>
            </span>
          </Metadata>
        </MetadataWrapper>
      </td>
    </MessageContentWrapper>
  );
};

export default MessageContent;
