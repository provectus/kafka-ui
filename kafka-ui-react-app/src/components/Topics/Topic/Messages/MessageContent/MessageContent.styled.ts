import styled, { css } from 'styled-components';
import * as SEditorViewer from 'components/common/EditorViewer/EditorViewer.styled';

export const Wrapper = styled.tr`
  background-color: ${({ theme }) => theme.topicMetaData.backgroundColor};
  & > td {
    padding: 16px;
    &:first-child {
      padding-right: 1px;
    }
    &:last-child {
      padding-left: 1px;
    }
  }
`;

export const Section = styled.div`
  padding: 0 16px;
  display: flex;
  gap: 1px;
  align-items: stretch;
`;

export const ContentBox = styled.div`
  background-color: ${({ theme }) => theme.topicMetaData.backgroundColor};
  padding: 24px;
  border-radius: 8px 0 0 8px;
  flex-grow: 3;
  display: flex;
  flex-direction: column;
  & nav {
    padding-bottom: 16px;
  }
  ${SEditorViewer.Wrapper} {
    flex-grow: 1;
  }
`;

export const MetadataWrapper = styled.div`
  background-color: ${({ theme }) => theme.topicMetaData.backgroundColor};
  padding: 24px;
  border-radius: 0 8px 8px 0;
  flex-grow: 1;
  display: flex;
  flex-direction: column;
  gap: 16px;
  min-width: 400px;
`;

export const Metadata = styled.span`
  display: flex;
  gap: 35px;
`;

export const MetadataLabel = styled.p`
  color: ${({ theme }) => theme.topicMetaData.color.label};
  font-size: 14px;
  width: 80px;
`;

export const MetadataValue = styled.p`
  color: ${({ theme }) => theme.topicMetaData.color.value};
  font-size: 14px;
`;

export const MetadataMeta = styled.p`
  color: ${({ theme }) => theme.topicMetaData.color.meta};
  font-size: 12px;
`;

export const Tab = styled.button<{ $active?: boolean }>(
  ({ theme, $active }) => css`
    background-color: ${theme.secondaryTab.backgroundColor[
      $active ? 'active' : 'normal'
    ]};
    color: ${theme.secondaryTab.color[$active ? 'active' : 'normal']};
    padding: 6px 16px;
    height: 32px;
    border: 1px solid ${theme.layout.stuffBorderColor};
    cursor: pointer;
    &:hover {
      background-color: ${theme.secondaryTab.backgroundColor.hover};
      color: ${theme.secondaryTab.color.hover};
    }
    &:first-child {
      border-radius: 4px 0 0 4px;
    }
    &:last-child {
      border-radius: 0 4px 4px 0;
    }
    &:not(:last-child) {
      border-right: 0;
    }
  `
);
export const Tabs = styled.nav``;
