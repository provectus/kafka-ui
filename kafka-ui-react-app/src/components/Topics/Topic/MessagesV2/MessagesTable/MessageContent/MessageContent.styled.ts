import styled, { css } from 'styled-components';
import * as SEditorViewer from 'components/common/EditorViewer/EditorViewer.styled';

export const Section = styled.div`
  display: grid;
  grid-template-columns: 1fr 400px;
  align-items: stretch;
`;

export const ContentBox = styled.div`
  background-color: white;
  border-right: 1px solid ${({ theme }) => theme.layout.stuffBorderColor};
  display: flex;
  flex-direction: column;
  padding-right: 16px;
  & nav {
    padding-bottom: 16px;
  }

  ${SEditorViewer.Wrapper} {
    flex-grow: 1;
  }
`;

export const MetadataWrapper = styled.div`
  padding-left: 16px;
`;

export const Tab = styled.button<{ $active?: boolean }>(
  ({ theme, $active }) => css`
    background-color: ${theme.secondaryTab.backgroundColor[
      $active ? 'active' : 'normal'
    ]};
    color: ${theme.secondaryTab.color[$active ? 'active' : 'normal']};
    padding: 6px 16px;
    height: 32px;
    border: 1px solid ${theme.layout.stuffBorderColor};
    cursor: pointer;
    &:hover {
      background-color: ${theme.secondaryTab.backgroundColor.hover};
      color: ${theme.secondaryTab.color.hover};
    }
    &:first-child {
      border-radius: 4px 0 0 4px;
    }
    &:last-child {
      border-radius: 0 4px 4px 0;
    }
    &:not(:last-child) {
      border-right: 0px;
    }
  `
);

export const Tabs = styled.nav``;
