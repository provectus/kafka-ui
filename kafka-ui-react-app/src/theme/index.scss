@import 'src/theme/bulma_overrides';

#root, body, html {
  width: 100%;
  position: relative;
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  background: repeating-linear-gradient(
    145deg,
    rgba(0,0,0,.003),
    rgba(0,0,0,.005) 5px,
    rgba(0,0,0,0) 5px,
    rgba(0,0,0,0) 10px
  ),
  repeating-linear-gradient(
    -145deg,
    rgba(0,0,0,.003),
    rgba(0,0,0,.005) 5px,
    rgba(0,0,0,0) 5px,
    rgba(0,0,0,0) 10px
  );
  background-color: $light;

}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
    monospace;
}

.cursor-pointer {
  cursor: pointer;
}
