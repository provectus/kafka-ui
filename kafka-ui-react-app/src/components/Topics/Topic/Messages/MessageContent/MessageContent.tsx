import React from 'react';
import EditorViewer from 'components/common/EditorViewer/EditorViewer';
import BytesFormatted from 'components/common/BytesFormatted/BytesFormatted';
import { SchemaType, TopicMessageTimestampTypeEnum } from 'generated-sources';
import { formatTimestamp } from 'lib/dateTimeHelpers';
import { useSearchParams } from 'react-router-dom';

import * as S from './MessageContent.styled';

type Tab = 'key' | 'content' | 'headers';

export interface MessageContentProps {
  messageKey?: string;
  messageContent?: string;
  headers?: { [key: string]: string | undefined };
  timestamp?: Date;
  timestampType?: TopicMessageTimestampTypeEnum;
}

const MessageContent: React.FC<MessageContentProps> = ({
  messageKey,
  messageContent,
  headers,
  timestamp,
  timestampType,
}) => {
  const [activeTab, setActiveTab] = React.useState<Tab>('content');
  const [searchParams] = useSearchParams();
  const keyFormat = searchParams.get('keySerde') || '';
  const valueFormat = searchParams.get('valueSerde') || '';

  const activeTabContent = () => {
    switch (activeTab) {
      case 'content':
        return messageContent;
      case 'key':
        return messageKey;
      default:
        return JSON.stringify(headers);
    }
  };

  const handleKeyTabClick = (e: React.MouseEvent) => {
    e.preventDefault();
    setActiveTab('key');
  };

  const handleContentTabClick = (e: React.MouseEvent) => {
    e.preventDefault();
    setActiveTab('content');
  };

  const handleHeadersTabClick = (e: React.MouseEvent) => {
    e.preventDefault();
    setActiveTab('headers');
  };
  const keySize = new TextEncoder().encode(messageKey).length;
  const contentSize = new TextEncoder().encode(messageContent).length;
  const contentType =
    messageContent && messageContent.trim().startsWith('{')
      ? SchemaType.JSON
      : SchemaType.PROTOBUF;

  return (
    <S.Wrapper>
      <td colSpan={10}>
        <S.Section>
          <S.ContentBox>
            <S.Tabs>
              <S.Tab
                type="button"
                $active={activeTab === 'key'}
                onClick={handleKeyTabClick}
              >
                Key
              </S.Tab>
              <S.Tab
                $active={activeTab === 'content'}
                type="button"
                onClick={handleContentTabClick}
              >
                Value
              </S.Tab>
              <S.Tab
                $active={activeTab === 'headers'}
                type="button"
                onClick={handleHeadersTabClick}
              >
                Headers
              </S.Tab>
            </S.Tabs>
            <EditorViewer
              data={activeTabContent() || ''}
              maxLines={28}
              schemaType={contentType}
            />
          </S.ContentBox>
          <S.MetadataWrapper>
            <S.Metadata>
              <S.MetadataLabel>Timestamp</S.MetadataLabel>
              <span>
                <S.MetadataValue>{formatTimestamp(timestamp)}</S.MetadataValue>
                <S.MetadataMeta>Timestamp type: {timestampType}</S.MetadataMeta>
              </span>
            </S.Metadata>

            <S.Metadata>
              <S.MetadataLabel>Key Serde</S.MetadataLabel>
              <span>
                <S.MetadataValue>{keyFormat}</S.MetadataValue>
                <S.MetadataMeta>
                  Size: <BytesFormatted value={keySize} />
                </S.MetadataMeta>
              </span>
            </S.Metadata>

            <S.Metadata>
              <S.MetadataLabel>Value Serde</S.MetadataLabel>
              <span>
                <S.MetadataValue>{valueFormat}</S.MetadataValue>
                <S.MetadataMeta>
                  Size: <BytesFormatted value={contentSize} />
                </S.MetadataMeta>
              </span>
            </S.Metadata>
          </S.MetadataWrapper>
        </S.Section>
      </td>
    </S.Wrapper>
  );
};

export default MessageContent;
