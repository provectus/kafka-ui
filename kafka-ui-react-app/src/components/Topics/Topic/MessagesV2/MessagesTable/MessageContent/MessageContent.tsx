import { SchemaType, TopicMessage } from 'generated-sources';
import React from 'react';
import EditorViewer from 'components/common/EditorViewer/EditorViewer';
import BytesFormatted from 'components/common/BytesFormatted/BytesFormatted';
import { formatTimestamp } from 'lib/dateTimeHelpers';
import { Row } from '@tanstack/react-table';
import {
  Label,
  List,
  SubText,
} from 'components/common/PropertiesList/PropertiesList.styled';

import * as S from './MessageContent.styled';

type Tab = 'key' | 'content' | 'headers';

const MessageContent: React.FC<{ row: Row<TopicMessage> }> = ({ row }) => {
  const {
    content,
    valueFormat,
    key,
    keyFormat,
    headers,
    timestamp,
    timestampType,
  } = row.original;

  const [activeTab, setActiveTab] = React.useState<Tab>('content');
  const activeTabContent = () => {
    switch (activeTab) {
      case 'content':
        return content;
      case 'key':
        return key;
      default:
        return JSON.stringify(headers || {});
    }
  };

  const encoder = new TextEncoder();
  const keySize = encoder.encode(key).length;
  const contentSize = encoder.encode(content).length;
  const contentType =
    content && content.trim().startsWith('{')
      ? SchemaType.JSON
      : SchemaType.PROTOBUF;
  return (
    <S.Section>
      <S.ContentBox>
        <S.Tabs>
          <S.Tab
            type="button"
            $active={activeTab === 'key'}
            onClick={() => setActiveTab('key')}
          >
            Key
          </S.Tab>
          <S.Tab
            $active={activeTab === 'content'}
            type="button"
            onClick={() => setActiveTab('content')}
          >
            Content
          </S.Tab>
          <S.Tab
            $active={activeTab === 'headers'}
            type="button"
            onClick={() => setActiveTab('headers')}
          >
            Headers
          </S.Tab>
        </S.Tabs>
        <EditorViewer
          data={activeTabContent() || ''}
          maxLines={28}
          schemaType={contentType}
        />
      </S.ContentBox>
      <S.MetadataWrapper>
        <List>
          <Label>Timestamp</Label>
          <span>
            {formatTimestamp(timestamp)}
            <SubText>Timestamp type: {timestampType}</SubText>
          </span>
          <Label>Content</Label>
          <span>
            {valueFormat}
            <SubText>
              Size: <BytesFormatted value={contentSize} />
            </SubText>
          </span>
          <Label>Key</Label>
          <span>
            {keyFormat}
            <SubText>
              Size: <BytesFormatted value={keySize} />
            </SubText>
          </span>
        </List>
      </S.MetadataWrapper>
    </S.Section>
  );
};

export default MessageContent;
