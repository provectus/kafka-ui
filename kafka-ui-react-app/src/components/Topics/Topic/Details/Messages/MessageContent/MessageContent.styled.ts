import styled from 'styled-components';
import { Colors } from 'theme/theme';

export const MessageContentWrapper = styled.tr`
  background-color: ${Colors.neutral[5]};
  & > td {
    padding: 16px;
    &:first-child {
      padding-right: 1px;
    }
    &:last-child {
      padding-left: 1px;
    }
  }
`;

export const StyledSection = styled.div`
  padding: 0 16px;
  display: flex;
  gap: 1px;
  align-items: stretch;
`;

export const ContentBox = styled.div`
  background-color: white;
  padding: 24px;
  border-radius: 8px 0 0 8px;
  flex-grow: 3;
  & nav {
    padding-bottom: 16px;
  }
`;

export const MetadataWrapper = styled.div`
  background-color: white;
  padding: 24px;
  border-radius: 0 8px 8px 0;
  flex-grow: 1;
  display: flex;
  flex-direction: column;
  gap: 16px;
  min-width: 400px;
`;

export const Metadata = styled.span`
  display: flex;
  gap: 16px;
`;

export const MetadataLabel = styled.p`
  color: ${Colors.neutral[50]};
  font-size: 14px;
  width: 80px;
`;

export const MetadataValue = styled.p`
  color: ${Colors.neutral[80]};
  font-size: 14px;
`;

export const MetadataMeta = styled.p`
  color: ${Colors.neutral[30]};
  font-size: 12px;
`;
