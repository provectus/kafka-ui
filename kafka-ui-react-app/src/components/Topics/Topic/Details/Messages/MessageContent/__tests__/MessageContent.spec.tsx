import { TextEncoder } from 'util';

import { fireEvent, render, waitFor } from '@testing-library/react';
import MessageContent, {
  MessageContentProps,
} from 'components/Topics/Topic/Details/Messages/MessageContent/MessageContent';
import { TopicMessageTimestampTypeEnum } from 'generated-sources';
import React from 'react';
import { ThemeProvider } from 'styled-components';
import theme from 'theme/theme';

const setupWrapper = (props?: Partial<MessageContentProps>) => {
  return (
    <ThemeProvider theme={theme}>
      <table>
        <tbody>
          <MessageContent
            messageKey='"test-key"'
            messageContent='{"data": "test"}'
            headers={{ header: 'test' }}
            timestamp={new Date(0)}
            timestampType={TopicMessageTimestampTypeEnum.CREATE_TIME}
            {...props}
          />
        </tbody>
      </table>
    </ThemeProvider>
  );
};

global.TextEncoder = TextEncoder;

describe('MessageContent component', () => {
  it('matches snapshot', () => {
    const component = render(setupWrapper());
    expect(component.baseElement).toMatchSnapshot();
  });

  describe('when switched to display the key', () => {
    it('has a tab with is-active classname', async () => {
      const component = render(setupWrapper());
      const keyTab = await component.findAllByText('Key');
      fireEvent.click(keyTab[0]);
      await waitFor(async () => {
        expect(keyTab[0]).toHaveClass('is-active');
      });
    });
    it('displays the key in the JSONViewer', async () => {
      const component = render(setupWrapper());
      const keyTab = await component.findAllByText('Key');
      fireEvent.click(keyTab[0]);
      const JSONViewer = await component.findByText('"test-key"');
      await waitFor(async () => {
        expect(JSONViewer).toBeTruthy();
      });
    });
  });

  describe('when switched to display the headers', () => {
    it('has a tab with is-active classname', async () => {
      const component = render(setupWrapper());
      const headersTab = await component.findByText('Headers');
      fireEvent.click(headersTab);
      await waitFor(async () => {
        expect(headersTab).toHaveClass('is-active');
      });
    });
    it('displays the key in the JSONViewer', async () => {
      const component = render(setupWrapper());
      const headersTab = await component.findByText('Headers');
      fireEvent.click(headersTab);
      const JSONViewer = await component.findByText('header:');
      await waitFor(async () => {
        expect(JSONViewer).toBeTruthy();
      });
    });
  });
});
