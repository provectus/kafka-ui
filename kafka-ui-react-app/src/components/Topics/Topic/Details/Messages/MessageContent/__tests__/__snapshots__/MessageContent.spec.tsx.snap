// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageContent component matches snapshot 1`] = `
<body>
  <div>
    <table>
      <tbody>
        <tr
          class="sc-gsTEea mhqOT"
        >
          <td
            colspan="5"
          >
            <div
              class="content-box content-wrapper"
            >
              <nav
                class="sc-bdfBQB kPaEqi"
              >
                <button
                  class=""
                  type="button"
                >
                  Key
                </button>
                <button
                  class="is-active"
                  type="button"
                >
                  Content
                </button>
                <button
                  class=""
                  type="button"
                >
                  Headers
                </button>
              </nav>
              <div
                class="json-viewer-wrapper"
              >
                <ul
                  style="border: 0px; padding: 0px; margin: 0.5em 0px 0.5em 0.125em; list-style: none; background-color: rgb(253, 255, 255);"
                >
                  <li
                    style="padding: 0px; margin: 0px;"
                  >
                    <ul
                      style="padding: 0px; margin: 0px; list-style: none;"
                    >
                      <li
                        style="padding-top: 0.25em; padding-right: 0px; margin-left: 0.875em; word-wrap: break-word; padding-left: 1.25em; text-indent: -0.5em; word-break: break-all;"
                      >
                        <label
                          style="display: inline-block; color: rgb(255, 83, 83); margin: 0px 0.5em 0px 0px;"
                        >
                          <span>
                            data
                            :
                          </span>
                        </label>
                        <span
                          style="color: rgb(36, 158, 77);"
                        >
                          "test"
                        </span>
                      </li>
                    </ul>
                  </li>
                </ul>
              </div>
            </div>
          </td>
          <td
            colspan="2"
          >
            <div
              class="content-box metadata-wrapper"
            >
              <span
                class="metadata"
              >
                <p
                  class="metadata-label"
                >
                  Timestamp
                </p>
                <span>
                  <p
                    class="metadata-value"
                  >
                    Thu, 01 Jan 1970 00:00:00 GMT
                  </p>
                  <p
                    class="metadata-meta"
                  >
                    Timestamp type: 
                    CREATE_TIME
                  </p>
                </span>
              </span>
              <span
                class="metadata"
              >
                <p
                  class="metadata-label"
                >
                  Content
                </p>
                <span>
                  <p
                    class="metadata-value"
                  >
                    JSON
                  </p>
                  <p
                    class="metadata-meta"
                  >
                    Size: 
                    <span>
                      16Bytes
                    </span>
                  </p>
                </span>
              </span>
              <span
                class="metadata"
              >
                <p
                  class="metadata-label"
                >
                  Key
                </p>
                <span>
                  <p
                    class="metadata-value"
                  >
                    Text
                  </p>
                  <p
                    class="metadata-meta"
                  >
                    Size: 
                    <span>
                      10Bytes
                    </span>
                  </p>
                </span>
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</body>
`;
